src/storage-controller/src/instance.rs

```diff
@@ -0,0 +1,444 @@
+    pub async fn recv(&mut self) -> Option<StorageResponse<T>> {
+        if let Some(resp) = self.ready_responses.pop_front() {
+            return Some(resp);
+        }
+
+        let receives = self.replicas.values_mut().map(|r| r.recv());
+        let mut futs: FuturesUnordered<_> = receives.collect();
+
+        if let Some(resp) = futs.next().await {
+            resp
+        } else {
+            // There are no live replicas left.
+            // Block forever to communicate that no response is ready.
+            futures::future::pending().await
+        }
+    }
+}
+
+/// Replica-specific configuration.
+#[derive(Clone, Debug)]
+pub(super) struct ReplicaConfig {
+    pub build_info: &'static BuildInfo,
+    pub location: ClusterReplicaLocation,
+    pub grpc_client: GrpcClientParameters,
+}
+
+/// State maintained about individual replicas.
+#[derive(Debug)]
+pub struct Replica<T> {
+    /// Replica configuration.
+    config: ReplicaConfig,
+    /// Whether the replica has failed and requires rehydration.
+    failed: bool,
+
+    /// A sender for commands for the replica.
+    ///
+    /// If sending to this channel fails, the replica has failed and requires
+    /// rehydration.
+    command_tx: mpsc::UnboundedSender<StorageCommand<T>>,
+    /// A receiver for responses from the replica.
+    ///
+    /// If receiving from the channel returns `None`, the replica has failed
+    /// and requires rehydration.
+    response_rx: mpsc::UnboundedReceiver<StorageResponse<T>>,
+    /// A handle to the task that aborts it when the replica is dropped.
+    _task: AbortOnDropHandle<()>,
+}
+
+impl<T> Replica<T>
+where
+    T: Timestamp + Lattice,
+    StorageGrpcClient: StorageClient<T>,
+{
+    /// Creates a new [`Replica`].
+    fn new(
+        id: ReplicaId,
+        config: ReplicaConfig,
+        epoch: ClusterStartupEpoch,
+        metrics: ReplicaMetrics,
+    ) -> Self {
+        let (command_tx, command_rx) = mpsc::unbounded_channel();
+        let (response_tx, response_rx) = mpsc::unbounded_channel();
+
+        let task = mz_ore::task::spawn(
+            || "storage-replica-{id}",
+            ReplicaTask {
+                replica_id: id,
+                config: config.clone(),
+                epoch,
+                metrics: metrics.clone(),
+                command_rx,
+                response_tx,
+            }
+            .run(),
+        );
+
+        Self {
+            config,
+            failed: false,
+            command_tx,
+            response_rx,
+            _task: task.abort_on_drop(),
+        }
+    }
+
+    /// Sends a command to the replica.
+    fn send(&mut self, command: StorageCommand<T>) {
+        if self.command_tx.send(command).is_err() {
+            self.failed = true;
+        }
+    }
+
+    /// Receives the next response from the replica and returns it, or `None` if the replica has
+    /// disconnected.
+    async fn recv(&mut self) -> Option<StorageResponse<T>> {
```

Comment:
We should mention in the documentation that this is a cancel-safe method and I would ideally write it manually like so:

```rust
/// Receives the next response from the replica and returns it, or `None` if the replica has
/// disconnected.
/// This method is cancel safe
fn recv(&mut self) -> impl Future<Output=Option<StorageResponse<T>> + '_ {
    std::future::poll_fn(|cx| {
        match std::task::ready!(self.response_tx.poll_recv(cx)) {
            Ok(resp) => Poll::Ready(Some(resp)),
            Err(_) => {
                self.failed = true;
                Poll::Ready(None)
            }
        }
    })
}
```

Reply:
I'll add the docs. It would be nice if one could somehow assert cancel-safety. It's still easy for people to not read the docs and break things.

Is there something about the manual rewrite that communicates cancel-safety better than the one using `await`?

Reply:
> Is there something about the manual rewrite that communicates cancel-safety better than the one using await?

Yes, a manual implementation makes it much clearer because cancel unsafety stems dropping a future after it had previously returned `Pending` which can leave data structures in some intermediary state or drop important local state held by the future and would later be used. By looking at the manual implementation you can verify that:
* Either `Poll::Pending` is returned or state is mutated but never both. If there was a case where state is mutated and then `Pending` was returned then we would have to reason about what happens if the execution stops there.
* No additional temporary state is held by the future. This is because all state must be captured by the closure and if it did need state it would have to explicitly create it with something like `let foo = vec![]` and then do `move |cx| { .. }` to move `foo` into the closure state.

Reply:
Thanks, with all this context it makes sense. But for most people, who don't have as much experience with async Rust, istm that the original version of this code is more obviously safe:
```rust
    async fn recv(&mut self) -> Option<StorageResponse<T>> {
        // `tokio::sync::mpsc::UnboundedReceiver::recv` is documented as cancel safe. 
        self.response_rx.recv().await.or_else(|| {
            self.failed = true;
            None
        })
    }
```

My naive reasoning would be: The method only contains a single `await` and that `await` is on a cancel safe method, so there is no opportunity for us to loose any state. Either the `await` returns a message, then we immediately pass it to the caller, or it doesn't and then no state is lost when we get aborted.

The non-manual version seems more obvious to me, not because the line of reasoning is simpler, but because it doesn't use low-level async constructs that I'd first have to look up to know what they are doing.

Reply:
> My naive reasoning would be: The method only contains a single `await` and that `await` is on a cancel safe method, so there is no opportunity for us to loose any state.

The problem with this is that state can be created accidentally and too easily. Contrived example but someone could insert a `let cur_state = std::mem::take(&mut self.failed);` line before awaiting the cancel-safe `UnboundedReceiver::recv`. You suddenly have a non-cancel safe method because state is mutated.

More subtle things can happen when a parameter is an expression that causes a temporary value to be created (the compiler inserts it) and that temporary has side-effects.

My take is that cancellation safety is subtle enough that the high bar of reasoning on the low level async machinery is required. I won't block merging on this but I strongly suggest that we switch to the [core register](https://without.boats/blog/the-registers-of-rust/).

Reply:
Fair enough. I still feel like the additional required context makes it harder to understand what's going on, but I guess the goal is not to make the method easy to understand but to make it hard to change in an incorrect way. Or rather, not breaking cancel safety is more important than readability.

I've rewritten the two cancel safe methods that are introduced by this PR. We can rewrite the others separately, although I'd still prefer if we could get rid of the cancel safety requirement instead :)